starboard/loader_app/slot_management.cc

```diff
@@ -148,9 +242,56 @@ void* LoadSlotManagedLibrary(const std::string& app_key,
     SB_LOG(WARNING) << "Failed to roll forward";
   }
 
+  int current_installation = ImGetCurrentInstallationIndex();
+
+  // Check the system image. If it's newer than the current slot, update to
+  // system image immediately.
+  if (current_installation != 0) {
+    std::vector<char> current_version(kMaxEgVersionLength);
+    std::vector<char> current_installation_path(kSbFileMaxPath);
+    if (ImGetInstallationPath(current_installation,
+                              current_installation_path.data(),
+                              kSbFileMaxPath) != IM_ERROR) {
+      std::vector<char> current_manifest_file_path(kSbFileMaxPath);
+      snprintf(current_manifest_file_path.data(), kSbFileMaxPath, "%s%s%s",
+               current_installation_path.data(), kSbFileSepString,
+               kManifestFileName);
+
+      if (!ReadEvergreenVersion(&current_manifest_file_path,
+                                current_version.data(), kMaxEgVersionLength)) {
+        SB_LOG(WARNING) << "Failed to read the Evergreen version of the "
+                           "current installation";
+      }
+    }
+
+    std::vector<char> system_image_version(kMaxEgVersionLength);
+    std::vector<char> system_image_installation_path(kSbFileMaxPath);
+    if (ImGetInstallationPath(0, system_image_installation_path.data(),
+                              kSbFileMaxPath) != IM_ERROR) {
+      std::vector<char> system_image_manifest_file_path(kSbFileMaxPath);
+      snprintf(system_image_manifest_file_path.data(), kSbFileMaxPath, "%s%s%s",
+               system_image_installation_path.data(), kSbFileSepString,
+               kManifestFileName);
+
+      if (!ReadEvergreenVersion(&system_image_manifest_file_path,
+                                system_image_version.data(),
+                                kMaxEgVersionLength)) {
+        SB_LOG(WARNING)
+            << "Failed to read the Evergreen version of the system image";
+      }
+    }
+
+    if (CompareEvergreenVersion(&system_image_version, &current_version) > 0) {
+      if (ImRollForward(0) != IM_ERROR) {
```

Comment:
It would be great if we could add a YTS test for this behavior. It seems difficult to do though since we don't have file system access from YTS (for example to update the system image to a new version, following an EG-powered update in an installation slot). Do you have any ideas for this?